src/misc/rwqueue.cpp

```diff
@@ -89,6 +136,52 @@ T RWQueue<T>::Dequeue()
 	return item;
 }
 
+template <typename T>
+void RWQueue<T>::BulkDequeue(std::vector<T>& into_target, const size_t num_requested)
+{
+	// Adjust the vector size, if needed
+	if (into_target.size() != num_requested) {
+		into_target.resize(num_requested);
+	}
+
+	// The target iterator will be incremented with each chunk dequeued
+	auto target_start = into_target.begin();
+
+	// The remaining count is decremented until no more items need dequeuing
+	auto num_remaining = num_requested;
+
+	while (num_remaining > 0) {
+		// Begin exclusive access to the internal queue
+		std::unique_lock<std::mutex> lock(mutex);
+
+		// Limit the quantity to our availability
+		const auto min_items = std::max(static_cast<size_t>(1), queue.size());
+		const auto num_items = std::min(num_remaining, min_items);
+		const auto stride    = static_cast<stride_t>(num_items);
+
+		// wait until the queue has enough items
+		has_items.wait(lock, [&] { return queue.size() >= num_items; });
+
+		// Set the range of the qeued items we'll be working with
+		const auto source_start = queue.begin();
+		const auto source_end   = source_start + stride;
+
+		// Dequeue the range of items into the target vector
+		std::move(source_start, source_end, target_start);
+		queue.erase(source_start, source_end);
+
+		// notify the first waiting thread that the queue now has room
+		lock.unlock();
+		has_room.notify_one();
+
+		// Increment to the next start forward
+		target_start += stride;
+
+		// Deduct the quantity done this pass
+		num_remaining -= num_items;
+	}
```

Comment:
My comments for `Enqueue` apply to this function as well:

- I think the logic could be simplified
- I'd get rid of most comments
- etc.